evm_loader/program/src/instruction/storage_to_v2/convert_data_account_from_v1_to_v2.rs

```diff
@@ -0,0 +1,273 @@
+use std::mem::size_of;
+
+use evm::U256;
+use solana_program::{msg, system_instruction};
+use solana_program::account_info::AccountInfo;
+use solana_program::entrypoint::ProgramResult;
+use solana_program::program::invoke;
+use solana_program::pubkey::Pubkey;
+use solana_program::rent::Rent;
+use solana_program::sysvar::Sysvar;
+
+use crate::account::{AccountData, ether_contract, Packable};
+use crate::account::ether_contract::DataV1;
+use crate::config::STORAGE_ENTIRIES_IN_CONTRACT_ACCOUNT;
+use crate::error::EvmLoaderError;
+use crate::hamt::Hamt;
+use crate::instruction::storage_to_v2::OPERATOR_PUBKEY;
+
+const TAG_SIZE: usize = size_of::<u8>();
+const STORAGE_ENTRY_SIZE: usize = size_of::<U256>();
+const CONTRACT_STORAGE_SIZE: usize =
+    STORAGE_ENTRY_SIZE * STORAGE_ENTIRIES_IN_CONTRACT_ACCOUNT as usize;
+
+enum AccountIndexes {
+    FundingAccount,
+    SystemProgram,
+    EthereumContract,
+}
+
+fn extract_contract_storage(hamt: &Hamt) -> Vec<u8> {
+    let mut contract_storage = vec![0u8; CONTRACT_STORAGE_SIZE];
+    for index in 0..STORAGE_ENTIRIES_IN_CONTRACT_ACCOUNT as usize {
+        if let Some(value) = hamt.find(index.into()) {
+            let offset = index * STORAGE_ENTRY_SIZE;
+            value.to_big_endian(&mut contract_storage[offset..offset + 32]);
+        }
+    }
+
+    contract_storage
+}
+
+fn convert_to_v2<'a>(
+    account_info: &'a AccountInfo<'a>,
+    funding_account: &'a AccountInfo<'a>,
+    system_program: &'a AccountInfo<'a>,
+) -> ProgramResult {
+    const DATA_END: usize = TAG_SIZE + DataV1::SIZE;
+    const GENERATION_FIELD_SIZE: usize = size_of::<u32>();
+
+    let data_size = account_info.data.borrow().len();
+
+    let (contract_storage, min_size_needed, addition_size, mut contract_storage_start) = {
+        let ethereum_contract_v1 =
+            AccountData::<ether_contract::DataV1, ether_contract::ExtensionV1>::from_account(
+                account_info.owner,
+                account_info,
+            )?;
+
+        let code_size = ethereum_contract_v1.code_size as usize;
+        let valids_size = code_size / 8 + 1;
+        let addition_size = code_size + valids_size;
+        let contract_storage_start = DATA_END + addition_size;
+
+        let min_size_needed = contract_storage_start + CONTRACT_STORAGE_SIZE + GENERATION_FIELD_SIZE;
+        if data_size < min_size_needed {
```

Comment:
What in case data_size > min_size_needed?We have some contracts which are bigger than needed.

Reply:
Fixed.